File: apps/desktop/src-tauri/src/updates/machine.rs
//! The desktop update state machine (design §2.1).
//!
//! It is deliberately free of Tauri, of the network and of the clock: every
//! transition is a pure function of the current state and one event, so the
//! table below is the whole contract and a test can walk all of it.
//!
//! Two rules matter more than the rest, and the transition function is written
//! so they cannot be broken by adding a case:
//!
//! 1. **Nothing becomes `UpToDate` except a check that came back saying so.**
//!    Not looking, failing to look, and looking and finding nothing are three
//!    different answers, and only the third one is "you are on the newest
//!    release".
//! 2. **An event a state does not accept changes nothing.** There is no
//!    fallthrough that lands somewhere plausible; an unexpected event is
//!    ignored and the state stays what it was.
//!
//! `Preparing` and `Installing` from the design's diagram are modelled as
//! phases of `Downloaded` rather than as separate reported states: what the
//! settings page and the tray have to say is still "downloaded, waiting for a
//! restart", and folding them in keeps the reported set to the eleven states of
//! design §4.1 while preserving every transition of §2.1.

use serde::{Deserialize, Serialize};

/// Why something could not be done, as a stable machine token.
///
/// Never a URL, a response body or a transport message: an endpoint can carry a
/// token, and a reason a person reads is rendered from this word by the front
/// end, in their language.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Reason {
    /// The release source could not be reached or read.
    SourceUnreachable,
    /// The source answered, but not with something this build can parse.
    SourceMalformed,
    /// A newer release exists and refuses this installed version.
    CompatibilityRefused,
    /// The release publishes nothing for this target.
    NoArtifactForTarget,
    /// The manifest entry is signed by a key this build does not carry.
    SignatureMismatch,
    /// The bytes that arrived are not the bytes the release described.
    DigestMismatch,
    /// The transfer stopped before it finished. Tauri cannot resume, so the
    /// bytes are discarded rather than kept as a half download.
    DownloadInterrupted,
    /// There was not enough room to stage the update.
    DiskFull,
    /// The desktop-owned Host would not stop, so the install never started.
    HostStopFailed,
    /// The installer itself failed; the running version was not replaced.
    InstallFailed,
    /// The updater plugin is present but unusable in this build.
    UpdaterUnavailable,
}

/// Which half of the updater configuration is still missing. Both are needed
/// before a check means anything: an endpoint says where releases are
/// published, and a public key is what makes one trustworthy.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MissingUpdaterConfig {
    pub pubkey: bool,
    pub endpoints: bool,
}

impl MissingUpdaterConfig {
    pub fn any(self) -> bool {
        self.pubkey || self.endpoints
    }
}

/// Why this shell cannot update itself at all, beyond configuration.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum Unsupported {
    /// The page is running in a browser, not in the desktop shell.
    NotDesktop,
    /// The Host answering is not on this machine, so replacing this shell would
    /// not be an update of anything the operator asked about.
    RemoteHost,
    /// The package manager that installed this build owns its updates.
    ManagedPackage,
}

/// What a release offers this target, once the Host's answer and the updater
/// manifest agree about it.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Offer {
    pub version: String,
    pub target: String,
    /// The `latest.json` of the same release, which is what the updater reads.
    pub manifest_url: String,
    /// The bundle the manifest points at, which is what actually gets applied.
    pub package_url: String,
    /// Lowercase hex, 64 characters, as published for `package_url`.
    pub sha256: String,
    pub size_bytes: u64,
    /// The release published a detached signature for this bundle. It is a
    /// property of the release, not a claim that anything was verified.
    pub signed: bool,
    pub notes_url: String,
}

/// Where in the restart a `Downloaded` update is.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum RestartPhase {
    /// Staged and idle; a person has not asked for the restart yet.
    #[default]
    Ready,
    /// Stopping the background this shell owns (design §2.3).
    Preparing,
    /// Handed to the installer. The process is expected to be replaced.
    Installing,
}

/// The eleven states the shell reports (design §4.1).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", tag = "state")]
pub enum UpdateState {
    /// No public key, no endpoint, or both. Nothing was consulted.
    NotConfigured {
        missing: MissingUpdaterConfig,
    },
    /// A build that never went through CI. It never auto-updates.
    LocalBuild,
    /// A shell that could not apply an update even if one existed.
    Unsupported {
        reason: Unsupported,
    },
    /// Configured, and nothing has been asked yet.
    Idle,
    Checking,
    UpToDate {
        checked_at_ms: i64,
    },
    /// The check was made and did not produce an answer. Never "up to date".
    Unavailable {
        reason: Reason,
        retry_after_ms: i64,
        checked_at_ms: i64,
    },
    Available {
        offer: Offer,
    },
    Downloading {
        offer: Offer,
        received_bytes: u64,
        total_bytes: u64,
    },
    Downloaded {
        offer: Offer,
        phase: RestartPhase,
        /// A restart that was started and did not get as far as installing.
        /// Kept beside the offer so the page can explain why it came back.
        problem: Option<Reason>,
    },
    Failed {
        reason: Reason,
        /// Retained so "retry" has something to retry. The bytes are gone; the
        /// offer is only a description of what to fetch again.
        offer: Option<Offer>,
    },
}

/// Everything that can happen to an update.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Event {
    /// A person pressed check, or the periodic timer fired.
    CheckStarted,
    /// A check in flight was abandoned.
    CheckCancelled,
    CheckedUpToDate {
        at_ms: i64,
    },
    CheckedAvailable {
        offer: Box<Offer>,
    },
    /// The Host or the manifest refused, with a reason and a retry hint.
    CheckRefused {
        reason: Reason,
        retry_after_ms: i64,
        at_ms: i64,
    },
    /// "Skip this version": the offer is dropped without claiming anything
    /// about whether a newer one exists.
    OfferDismissed,
    DownloadStarted,
    DownloadProgressed {
        received_bytes: u64,
        total_bytes: u64,
    },
    /// Signature and digest both passed; the bytes are on disk.
    DownloadFinished,
    DownloadFailed {
        reason: Reason,
    },
    /// A person confirmed "restart and update".
    RestartRequested,
    /// The desktop-owned background is stopped; the installer may run.
    BackgroundStopped,
    /// Stopping failed, or the person cancelled at the confirmation.
    RestartAbandoned {
        reason: Option<Reason>,
    },
    InstallFailed {
        reason: Reason,
    },
    /// "Try again" from a failure.
    Retry,
}

/// The state plus the transition table. Holding it rather than a bare enum
/// keeps `apply` the only way the state changes.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Machine {
    state: UpdateState,
}

impl Machine {
    /// A shell that has everything it needs and has not been asked yet.
    pub fn idle() -> Self {
        Self {
            state: UpdateState::Idle,
        }
    }

    /// A shell that was never given a key or an endpoint.
    pub fn not_configured(missing: MissingUpdaterConfig) -> Self {
        Self {
            state: UpdateState::NotConfigured { missing },
        }
    }

    pub fn local_build() -> Self {
        Self {
            state: UpdateState::LocalBuild,
        }
    }

    pub fn unsupported(reason: Unsupported) -> Self {
        Self {
            state: UpdateState::Unsupported { reason },
        }
    }

    pub fn state(&self) -> &UpdateState {
        &self.state
    }

    /// True while nothing about this build can ever change: these three states
    /// describe the build, not a check, so no event moves out of them.
    pub fn is_terminal(&self) -> bool {
        matches!(
            self.state,
            UpdateState::NotConfigured { .. }
                | UpdateState::LocalBuild
                | UpdateState::Unsupported { .. }
        )
    }

    /// Whether a check may start right now. A download or a pending restart is
    /// not interrupted by one: the offer in hand is what the person acted on.
    pub fn may_check(&self) -> bool {
        matches!(
            self.state,
            UpdateState::Idle
                | UpdateState::UpToDate { .. }
                | UpdateState::Unavailable { .. }
                | UpdateState::Available { .. }
                | UpdateState::Failed { .. }
        )
    }

    /// The offer this state is about, if any.
    pub fn offer(&self) -> Option<&Offer> {
        match &self.state {
            UpdateState::Available { offer }
            | UpdateState::Downloading { offer, .. }
            | UpdateState::Downloaded { offer, .. } => Some(offer),
            UpdateState::Failed { offer, .. } => offer.as_ref(),
            _ => None,
        }
    }

    /// Applies one event. Returns whether the state changed.
    ///
    /// Every arm is written out; there is no catch-all that guesses. An event
    /// arriving in a state that does not list it is a race the caller lost —
    /// a progress callback landing after a cancel, say — and losing a race must
    /// never invent an answer.
    pub fn apply(&mut self, event: Event) -> bool {
        let next = self.next(event);
        match next {
            Some(state) if state != self.state => {
                self.state = state;
                true
            }
            _ => false,
        }
    }

    fn next(&self, event: Event) -> Option<UpdateState> {
        if self.is_terminal() {
            return None;
        }
        match (&self.state, event) {
            // ---- checking -------------------------------------------------
            (_, Event::CheckStarted) if self.may_check() => Some(UpdateState::Checking),
            (UpdateState::Checking, Event::CheckCancelled) => Some(UpdateState::Idle),
            (UpdateState::Checking, Event::CheckedUpToDate { at_ms }) => {
                Some(UpdateState::UpToDate {
                    checked_at_ms: at_ms,
                })
            }
            (UpdateState::Checking, Event::CheckedAvailable { offer }) => {
                Some(UpdateState::Available { offer: *offer })
            }
            (
                UpdateState::Checking,
                Event::CheckRefused {
                    reason,
                    retry_after_ms,
                    at_ms,
                },
            ) => Some(UpdateState::Unavailable {
                reason,
                retry_after_ms: retry_after_ms.max(0),
                checked_at_ms: at_ms,
            }),

            // ---- the offer ------------------------------------------------
            (UpdateState::Available { .. }, Event::OfferDismissed) => Some(UpdateState::Idle),
            (UpdateState::Available { offer }, Event::DownloadStarted) => {
                Some(UpdateState::Downloading {
                    offer: offer.clone(),
                    received_bytes: 0,
                    total_bytes: offer.size_bytes,
                })
            }
            (
                UpdateState::Failed {
                    offer: Some(offer), ..
                },
                Event::Retry,
            ) => Some(UpdateState::Available {
                offer: offer.clone(),
            }),
            (UpdateState::Failed { offer: None, .. }, Event::Retry) => Some(UpdateState::Idle),

            // ---- the transfer ---------------------------------------------
            (
                UpdateState::Downloading { offer, .. },
                Event::DownloadProgressed {
                    received_bytes,
                    total_bytes,
                },
            ) => Some(UpdateState::Downloading {
                offer: offer.clone(),
                received_bytes,
                // A server that reports no length leaves the offer's own size
                // as the best number there is; zero would render as a finished
                // bar the moment anything arrived.
                total_bytes: if total_bytes == 0 {
                    offer.size_bytes
                } else {
                    total_bytes
                },
            }),
            (UpdateState::Downloading { offer, .. }, Event::DownloadFinished) => {
                Some(UpdateState::Downloaded {
                    offer: offer.clone(),
                    phase: RestartPhase::Ready,
                    problem: None,
                })
            }
            (UpdateState::Downloading { offer, .. }, Event::DownloadFailed { reason }) => {
                Some(UpdateState::Failed {
                    reason,
                    offer: Some(offer.clone()),
                })
            }
            (UpdateState::Downloading { offer, .. }, Event::CheckCancelled) => {
                // Cancelling a download discards the bytes and keeps the offer:
                // a partial file is not something to resume from.
                Some(UpdateState::Available {
                    offer: offer.clone(),
                })
            }

            // ---- the restart ----------------------------------------------
            (
                UpdateState::Downloaded {
                    offer,
                    phase: RestartPhase::Ready,
                    ..
                },
                Event::RestartRequested,
            ) => Some(UpdateState::Downloaded {
                offer: offer.clone(),
                phase: RestartPhase::Preparing,
                problem: None,
            }),
            (
                UpdateState::Downloaded {
                    offer,
                    phase: RestartPhase::Preparing,
                    ..
                },
                Event::BackgroundStopped,
            ) => Some(UpdateState::Downloaded {
                offer: offer.clone(),
                phase: RestartPhase::Installing,
                problem: None,
            }),
            (
                UpdateState::Downloaded {
                    offer,
                    phase: RestartPhase::Preparing,
                    ..
                },
                Event::RestartAbandoned { reason },
            ) => Some(UpdateState::Downloaded {
                offer: offer.clone(),
                phase: RestartPhase::Ready,
                problem: reason,
            }),
            (
                UpdateState::Downloaded {
                    offer,
                    phase: RestartPhase::Installing,
                    ..
                },
                Event::InstallFailed { reason },
            ) => Some(UpdateState::Failed {
                reason,
                offer: Some(offer.clone()),
            }),

            _ => None,
        }
    }
}

impl Default for Machine {
    fn default() -> Self {
        Self::idle()
    }
}

File: apps/desktop/src/shell-core/updates/machine.ts
/**
 * The desktop update state machine (design §2.1).
 *
 * A line-by-line port of `src-tauri/src/updates/machine.rs`. It is deliberately
 * free of Electron, of the network and of the clock: every transition is a pure
 * function of the current state and one event, so the table below is the whole
 * contract and a test can walk all of it.
 *
 * Two rules matter more than the rest, and the transition function is written
 * so they cannot be broken by adding a case:
 *
 * 1. **Nothing becomes `upToDate` except a check that came back saying so.**
 *    Not looking, failing to look, and looking and finding nothing are three
 *    different answers, and only the third one is "you are on the newest
 *    release".
 * 2. **An event a state does not accept changes nothing.** There is no
 *    fallthrough that lands somewhere plausible; an unexpected event is
 *    ignored and the state stays what it was.
 *
 * `preparing` and `installing` from the design's diagram are modelled as
 * phases of `downloaded` rather than as separate reported states: what the
 * settings page and the tray have to say is still "downloaded, waiting for a
 * restart", and folding them in keeps the reported set to the eleven states of
 * design §4.1 while preserving every transition of §2.1.
 *
 * The wire names — every state tag, every reason token, every field — are the
 * ones `apps/web/src/updates/shell-updater.ts:17-62` already reads. They are
 * not an implementation detail of this file; a rename here is a silently broken
 * settings page.
 */

/**
 * Why something could not be done, as a stable machine token.
 *
 * Never a URL, a response body or a transport message: an endpoint can carry a
 * token, and a reason a person reads is rendered from this word by the front
 * end, in their language.
 */
export const REASONS = [
  /** The release source could not be reached or read. */
  "sourceUnreachable",
  /** The source answered, but not with something this build can parse. */
  "sourceMalformed",
  /** A newer release exists and refuses this installed version. */
  "compatibilityRefused",
  /** The release publishes nothing for this target. */
  "noArtifactForTarget",
  /** The manifest entry is signed by a key this build does not carry. */
  "signatureMismatch",
  /** The bytes that arrived are not the bytes the release described. */
  "digestMismatch",
  /**
   * The transfer stopped before it finished. The bytes are discarded rather
   * than kept as a half download.
   */
  "downloadInterrupted",
  /** There was not enough room to stage the update. */
  "diskFull",
  /** The desktop-owned Host would not stop, so the install never started. */
  "hostStopFailed",
  /** The installer itself failed; the running version was not replaced. */
  "installFailed",
  /** The updater is present but unusable in this build. */
  "updaterUnavailable",
] as const;

export type Reason = (typeof REASONS)[number];

/**
 * Which half of the updater configuration is still missing. Both are needed
 * before a check means anything: an endpoint says where releases are
 * published, and — under Tauri — a public key is what made one trustworthy.
 *
 * Under electron-updater the `pubkey` half carries the same meaning one step
 * over: it is true when this build carries nothing that could make an update
 * trustworthy, which is now "the package is not signed" rather than "no
 * minisign key was configured". The name is kept because it is the wire format
 * the settings page renders (`updates.missing.pubkey`).
 */
export interface MissingUpdaterConfig {
  pubkey: boolean;
  endpoints: boolean;
}

export function anyMissing(missing: MissingUpdaterConfig): boolean {
  return missing.pubkey || missing.endpoints;
}

/** Why this shell cannot update itself at all, beyond configuration. */
export type Unsupported =
  /** The page is running in a browser, not in the desktop shell. */
  | "notDesktop"
  /**
   * The Host answering is not on this machine, so replacing this shell would
   * not be an update of anything the operator asked about.
   */
  | "remoteHost"
  /** The package manager that installed this build owns its updates. */
  | "managedPackage";

/**
 * What a release offers this target, once the Host's answer and the release
 * manifest agree about it.
 */
export interface Offer {
  version: string;
  target: string;
  /** The manifest of the same release, which is what the updater reads. */
  manifestUrl: string;
  /** The bundle the manifest points at, which is what actually gets applied. */
  packageUrl: string;
  /** Lowercase hex, 64 characters, as published for `packageUrl`. */
  sha256: string;
  sizeBytes: number;
  /**
   * The release published a signature for this bundle. It is a property of the
   * release, not a claim that anything was verified.
   */
  signed: boolean;
  notesUrl: string;
}

/** Where in the restart a `downloaded` update is. */
export type RestartPhase =
  /** Staged and idle; a person has not asked for the restart yet. */
  | "ready"
  /** Stopping the background this shell owns (design §2.3). */
  | "preparing"
  /** Handed to the installer. The process is expected to be replaced. */
  | "installing";

/** The eleven states the shell reports (design §4.1). */
export type UpdateState =
  /** Nothing trustworthy, no endpoint, or both. Nothing was consulted. */
  | { state: "notConfigured"; missing: MissingUpdaterConfig }
  /** A build that never went through CI. It never auto-updates. */
  | { state: "localBuild" }
  /** A shell that could not apply an update even if one existed. */
  | { state: "unsupported"; reason: Unsupported }
  /** Configured, and nothing has been asked yet. */
  | { state: "idle" }
  | { state: "checking" }
  | { state: "upToDate"; checkedAtMs: number }
  /** The check was made and did not produce an answer. Never "up to date". */
  | {
      state: "unavailable";
      reason: Reason;
      retryAfterMs: number;
      checkedAtMs: number;
    }
  | { state: "available"; offer: Offer }
  | {
      state: "downloading";
      offer: Offer;
      receivedBytes: number;
      totalBytes: number;
    }
  | {
      state: "downloaded";
      offer: Offer;
      phase: RestartPhase;
      /**
       * A restart that was started and did not get as far as installing. Kept
       * beside the offer so the page can explain why it came back.
       */
      problem: Reason | null;
    }
  | {
      state: "failed";
      reason: Reason;
      /**
       * Retained so "retry" has something to retry. The bytes are gone; the
       * offer is only a description of what to fetch again.
       */
      offer: Offer | null;
    };

/** Everything that can happen to an update. */
export type Event =
  /** A person pressed check, or the periodic timer fired. */
  | { type: "checkStarted" }
  /** A check in flight was abandoned. */
  | { type: "checkCancelled" }
  | { type: "checkedUpToDate"; atMs: number }
  | { type: "checkedAvailable"; offer: Offer }
  /** The Host or the manifest refused, with a reason and a retry hint. */
  | {
      type: "checkRefused";
      reason: Reason;
      retryAfterMs: number;
      atMs: number;
    }
  /**
   * "Skip this version": the offer is dropped without claiming anything about
   * whether a newer one exists.
   */
  | { type: "offerDismissed" }
  | { type: "downloadStarted" }
  | { type: "downloadProgressed"; receivedBytes: number; totalBytes: number }
  /** Signature and digest both passed; the bytes are on disk. */
  | { type: "downloadFinished" }
  | { type: "downloadFailed"; reason: Reason }
  /** A person confirmed "restart and update". */
  | { type: "restartRequested" }
  /** The desktop-owned background is stopped; the installer may run. */
  | { type: "backgroundStopped" }
  /** Stopping failed, or the person cancelled at the confirmation. */
  | { type: "restartAbandoned"; reason: Reason | null }
  | { type: "installFailed"; reason: Reason }
  /** "Try again" from a failure. */
  | { type: "retry" };

/**
 * Structural equality over the plain data above. `apply` reports whether the
 * state changed, and "changed" has to mean the value differs — not that a new
 * object was allocated, which is true of every branch below.
 */
export function sameState(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (a === null || b === null) return false;
  if (typeof a !== "object" || typeof b !== "object") return false;
  const left = a as Record<string, unknown>;
  const right = b as Record<string, unknown>;
  const keys = Object.keys(left);
  if (keys.length !== Object.keys(right).length) return false;
  return keys.every(
    (key) =>
      Object.prototype.hasOwnProperty.call(right, key) &&
      sameState(left[key], right[key]),
  );
}

/**
 * The state plus the transition table. Holding it rather than a bare union
 * keeps `apply` the only way the state changes.
 */
export class Machine {
  private current: UpdateState;

  private constructor(state: UpdateState) {
    this.current = state;
  }

  /** A shell that has everything it needs and has not been asked yet. */
  static idle(): Machine {
    return new Machine({ state: "idle" });
  }

  /** A shell that was never given anything to trust, or anywhere to look. */
  static notConfigured(missing: MissingUpdaterConfig): Machine {
    return new Machine({ state: "notConfigured", missing: { ...missing } });
  }

  static localBuild(): Machine {
    return new Machine({ state: "localBuild" });
  }

  static unsupported(reason: Unsupported): Machine {
    return new Machine({ state: "unsupported", reason });
  }

  /** A machine already in a reported state, for the shell's own bootstrap. */
  static of(state: UpdateState): Machine {
    return new Machine(state);
  }

  state(): UpdateState {
    return this.current;
  }

  clone(): Machine {
    return new Machine(structuredClone(this.current));
  }

  /**
   * True while nothing about this build can ever change: these three states
   * describe the build, not a check, so no event moves out of them.
   */
  isTerminal(): boolean {
    return (
      this.current.state === "notConfigured" ||
      this.current.state === "localBuild" ||
      this.current.state === "unsupported"
    );
  }

  /**
   * Whether a check may start right now. A download or a pending restart is
   * not interrupted by one: the offer in hand is what the person acted on.
   */
  mayCheck(): boolean {
    switch (this.current.state) {
      case "idle":
      case "upToDate":
      case "unavailable":
      case "available":
      case "failed":
        return true;
      default:
        return false;
    }
  }

  /** The offer this state is about, if any. */
  offer(): Offer | null {
    switch (this.current.state) {
      case "available":
      case "downloading":
      case "downloaded":
        return this.current.offer;
      case "failed":
        return this.current.offer;
      default:
        return null;
    }
  }

  /**
   * Applies one event. Returns whether the state changed.
   *
   * Every arm is written out; there is no catch-all that guesses. An event
   * arriving in a state that does not list it is a race the caller lost — a
   * progress callback landing after a cancel, say — and losing a race must
   * never invent an answer.
   */
  apply(event: Event): boolean {
    const next = this.next(event);
    if (next === null || sameState(next, this.current)) return false;
    this.current = next;
    return true;
  }

  private next(event: Event): UpdateState | null {
    if (this.isTerminal()) return null;
    const state = this.current;

    // ---- checking ---------------------------------------------------------
    if (event.type === "checkStarted") {
      return this.mayCheck() ? { state: "checking" } : null;
    }
    if (state.state === "checking") {
      switch (event.type) {
        case "checkCancelled":
          return { state: "idle" };
        case "checkedUpToDate":
          return { state: "upToDate", checkedAtMs: event.atMs };
        case "checkedAvailable":
          return { state: "available", offer: event.offer };
        case "checkRefused":
          return {
            state: "unavailable",
            reason: event.reason,
            retryAfterMs: Math.max(event.retryAfterMs, 0),
            checkedAtMs: event.atMs,
          };
        default:
          return null;
      }
    }

    // ---- the offer --------------------------------------------------------
    if (state.state === "available") {
      switch (event.type) {
        case "offerDismissed":
          return { state: "idle" };
        case "downloadStarted":
          return {
            state: "downloading",
            offer: state.offer,
            receivedBytes: 0,
            totalBytes: state.offer.sizeBytes,
          };
        default:
          return null;
      }
    }
    if (state.state === "failed" && event.type === "retry") {
      return state.offer === null
        ? { state: "idle" }
        : { state: "available", offer: state.offer };
    }

    // ---- the transfer -----------------------------------------------------
    if (state.state === "downloading") {
      switch (event.type) {
        case "downloadProgressed":
          return {
            state: "downloading",
            offer: state.offer,
            receivedBytes: event.receivedBytes,
            // A server that reports no length leaves the offer's own size as
            // the best number there is; zero would render as a finished bar
            // the moment anything arrived.
            totalBytes:
              event.totalBytes === 0 ? state.offer.sizeBytes : event.totalBytes,
          };
        case "downloadFinished":
          return {
            state: "downloaded",
            offer: state.offer,
            phase: "ready",
            problem: null,
          };
        case "downloadFailed":
          return { state: "failed", reason: event.reason, offer: state.offer };
        case "checkCancelled":
          // Cancelling a download discards the bytes and keeps the offer: a
          // partial file is not something to resume from.
          return { state: "available", offer: state.offer };
        default:
          return null;
      }
    }

    // ---- the restart ------------------------------------------------------
    if (state.state === "downloaded") {
      if (state.phase === "ready" && event.type === "restartRequested") {
        return {
          state: "downloaded",
          offer: state.offer,
          phase: "preparing",
          problem: null,
        };
      }
      if (state.phase === "preparing" && event.type === "backgroundStopped") {
        return {
          state: "downloaded",
          offer: state.offer,
          phase: "installing",
          problem: null,
        };
      }
      if (state.phase === "preparing" && event.type === "restartAbandoned") {
        return {
          state: "downloaded",
          offer: state.offer,
          phase: "ready",
          problem: event.reason,
        };
      }
      if (state.phase === "installing" && event.type === "installFailed") {
        return { state: "failed", reason: event.reason, offer: state.offer };
      }
      return null;
    }

    return null;
  }
}
